### DISCUSSION TITLE
Sorting files numerically

### DISCUSSION THREAD

**Discussion by zefr0x** [nvim-tree/nvim-tree.lua#1896] [Ideas]
**Is your feature request related to a problem? Please describe.**
I'm always frustrated when I have a directory that has number-named files.
For example NvimTree will display them like that:
<details>

```
10
11
12
13
14
15
16
17
18
19
'1 hello'
'1 hello 4'
2
20
'3 hello'
'3 hello 5'
6
7
8
9
'hello 3'
```

</details>

But the normal order I want them to be in is like:
<details>

```
'1 hello'
'1 hello 4'
2
'3 hello'
'3 hello 5'
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
'hello 3'

```

</details>

It's the same default behavior of the `ls` GUN utility in Linux, but the `ls` has an option `-v` to treat filename numerically. https://unix.stackexchange.com/questions/33909/list-files-sorted-numerically

Most times I need numbered files sorted with the numerically normal order so I can walk between them easily.

**Describe the solution you'd like**
Maybe adding a new type to the `sort_by` option to do that type of sorting.

**Describe alternatives you've considered**
Maybe I just can create a function to do that since `sort_by` option support passing a function.

**Comment by alex-courtis** [comment]
You can add this functionality yourself. See `:help nvim-tree.sort_by`

This might get you started:
```lua
  local sort_by = function(nodes)
    table.sort(nodes, function(a, b)
      local a_num = string.match(a.name, "^[0-9]+")
      local b_num = string.match(b.name, "^[0-9]+")
      if a_num and b_num then
        return a_num < b_num
      else
        -- we need to do something here
        return false
      end
    end)
  end
```

**Comment by zefr0x** [comment]
I'm not even a beginner in Lua, but I made this:
```lua
local compare = function(left, right)
	left = left.name:lower()
	right = right.name:lower()

	for i = 1, string.len(left), 1 do
		local l = string.sub(left, i, -1)
		local r = string.sub(right, i, -1)

		if l == r then
			return false
		elseif type(tonumber(string.sub(l, 1, 1))) == "number" and type(tonumber(string.sub(r, 1, 1))) == "number" then
			return tonumber(string.match(l, "^[0-9]+")) < tonumber(string.match(r, "^[0-9]+"))
		elseif l[1] == r[1] then
			goto continue
		else
			return l < r
		end
		::continue::
	end
end


-- nvim-tree.lua
require("nvim-tree").setup({
	sort_by = function(nodes)
		table.sort(nodes, compare)
	end,
})

```

It seems to work fine. If any one have any optimizations or fail points please tell me.

References:

https://github.com/lifthrasiir/rust-natord/blob/93e5f0ace208e73ddfaaa77ec1605584c8f22a14/lib.rs#L142

https://github.com/lifthrasiir/rust-natord/blob/93e5f0ace208e73ddfaaa77ec1605584c8f22a14/lib.rs#L37

**Reply by alex-courtis** [reply]
That's fantastic!

It would be great if you added this [Recipe](https://github.com/nvim-tree/nvim-tree.lua/wiki/Recipes) to the wiki.

**Reply by zefr0x** [reply]
> It would be great if you added this [Recipe](https://github.com/nvim-tree/nvim-tree.lua/wiki/Recipes) to the wiki.

Done.

**Reply by admbeck** [reply]
This sort produces inconsistent results for me. After each write of the file nvim-tree re-sorts itself and fails to sort files and directories properly. I am not sure why it happens
![Screenshot_20230221_093804](https://user-images.githubusercontent.com/50273349/220249082-79ce35f6-0db1-43d7-8921-cd8df0c93fc8.png)

**Reply by zefr0x** [reply]
I think that this work:
```diff
local compare = function(left, right)
	left = left.name:lower()
	right = right.name:lower()

	for i = 1, string.len(left), 1 do
		local l = string.sub(left, i, -1)
		local r = string.sub(right, i, -1)

		if l == r then
			return false
		elseif type(tonumber(string.sub(l, 1, 1))) == "number" and type(tonumber(string.sub(r, 1, 1))) == "number" then
			return tonumber(string.match(l, "^[0-9]+")) < tonumber(string.match(r, "^[0-9]+"))
-		elseif l[1] == r[1] then
-			goto continue
		else
			return l < r
		end
-		::continue::
	end
end
```

My subconscious mind told me to get the `else` case out of the for loop, and it worked after this change.

It appears that this `else` case was not reachable, since the `l` and `r` are not indexable and the result will be `nil`, so this statement will be always true. Anyway it was a duplicate statement since the `l == r` is checked in the first statement.

Now, numbers sorting will work only if the number is in the start of the name. For now I'm trying to make it work for number if they are in the middle of end of the name.

Sorry about that, but my `lua` knowledge is not great, so my code is messy.

If you found any problems just say 👍

**Reply by zefr0x** [reply]
This is it, the fixed and improved version:

```lua
local function natural_cmp(left, right)
	left = left.name:lower()
	right = right.name:lower()

	if left == right then
		return false
	end

	for i = 1, math.max(string.len(left), string.len(right)), 1 do
		local l = string.sub(left, i, -1)
		local r = string.sub(right, i, -1)

		if type(tonumber(string.sub(l, 1, 1))) == "number" and type(tonumber(string.sub(r, 1, 1))) == "number" then
			local l_number = tonumber(string.match(l, "^[0-9]+"))
			local r_number = tonumber(string.match(r, "^[0-9]+"))

			if l_number ~= r_number then
				return l_number < r_number
			end
		elseif string.sub(l, 1, 1) ~= string.sub(r, 1, 1) then
			return l < r
		end
	end
end

require("nvim-tree").setup({
	sort_by = function(nodes)
		table.sort(nodes, natural_cmp)
	end,
})
```

**Reply by alex-courtis** [reply]
Thanks for the update!

**Comment by russarr** [comment]
Hi. I have a problem.
Sorting mixes files and directories. Is there a way to make directories at the top of the list as usual.

**Reply by alex-courtis** [reply]
> Hi. I have a problem. Sorting mixes files and directories. Is there a way to make directories at the top of the list as usual.

In your `:help nvim-tree.sort_by` function you can compare first on:
```
type:          "directory" | "file" | "link"
```

**Reply by russarr** [reply]
I don't understand how to use this parameter with a function from this topic
`require("nvim-tree").setup({
	sort_by = function(nodes)
		table.sort(nodes, natural_cmp)
	end,
})`
How should I specify the directory parameter.

**Reply by alex-courtis** [reply]
> How should I specify the directory parameter.

Extending the example in `:help nvim-tree.sort_by` to sort by length of name with directories first:
```lua
local sort_by = function(nodes)
  table.sort(nodes, function(a, b)
    -- directories first
    if a.type ~= "directory" and b.type == "directory" then
      return false
    elseif a.type == "directory" and b.type ~= "directory" then
      return true
    end

    -- sort by name length
    return #a.name < #b.name
  end)
end
```

**Reply by alex-courtis** [reply]
Refactored as per the above example with the comparison function:
```lua
local natural_cmp = function(left, right)
  -- directories first
  if left.type ~= "directory" and right.type == "directory" then
---
```

**Comment by russarr** [comment]
I need help again. I'm trying to merge two recipes:
 - https://github.com/nvim-tree/nvim-tree.lua/wiki/Recipes#sorting-files-naturally-respecting-numbers-within-files-names 
 - https://github.com/nvim-tree/nvim-tree.lua/wiki/Recipes#cycle-sort-methods
Individually, they work perfectly
For example, this works:
```
    require("nvim-tree").setup({
        sort_by = sort_by_natural,
    })
```
But if I try to add a sorting function to the list for cycling from second reciept, it doesn't work. Cycling is working, but natural sort doesn't. It just sorting by "name".
```
    local SORT_METHODS = {
      sort_by_natural,
      "case_sensitive",
      "modification_time",
      "extension",
    }
...
    require("nvim-tree").setup({
        sort_by = sort_by,
    })
```

**Reply by alex-courtis** [reply]
A `:help nvim-tree.sort.sorter` may return either a string with the predefined sorter or do the actual sort.

This synthetic example might help:

```lua
local sort_by_len = true

local sorter = function(nodes)
  if sort_by_len then
    table.sort(nodes, function(a, b)
      return #a.name < #b.name
    end)
  else
    return "modification_time"
  end
end

local cycle_sort = function()
  sort_by_len = not sort_by_len
  api.tree.reload()
end

---

  vim.keymap.set('n', 'T', cycle_sort, opts('Cycle Sort'))
```

**Comment by russarr** [comment]
Thanks!
So this in my resulting sort functions. Difference from original reciepts:
 - when sorting by name with the natural order of numbers, the directories remain on top
 - cycling between "natural names" and "modification_time"
```
    local sort_by_name = true
    local function sort_by_natural(nodes)
      local function sorter(left, right)
        if left.type ~= "directory" and right.type == "directory" then
          return false
        elseif left.type == "directory" and right.type ~= "directory" then
          return true
        end
        left = left.name:lower()
        right = right.name:lower()

        if left == right then
          return false
        end

        for i = 1, math.max(string.len(left), string.len(right)), 1 do
          local l = string.sub(left, i, -1)
          local r = string.sub(right, i, -1)

          if
              type(tonumber(string.sub(l, 1, 1))) == "number"
              and type(tonumber(string.sub(r, 1, 1))) == "number"
          then
            local l_number = tonumber(string.match(l, "^[0-9]+"))
            local r_number = tonumber(string.match(r, "^[0-9]+"))

            if l_number ~= r_number then
              return l_number < r_number
            end
          elseif string.sub(l, 1, 1) ~= string.sub(r, 1, 1) then
            return l < r
          end
        end
      end

      if sort_by_name then
        table.sort(nodes, sorter)
      else
        return "modification_time"
      end
    end

    local cycle_sort = function()
      sort_by_name = not sort_by_name
      api.tree.reload()
      if sort_by_name then
        vim.notify("Sorting by name")
      else
        vim.notify("Sorting by modification time")
      end
    end
---
vim.keymap.set('n', 'T', cycle_sort, opts('Cycle Sort'))

require("nvim-tree").setup({
     sort_by = sort_by_natural,
...
})
```

**Reply by alex-courtis** [reply]
Very nice. You can always add more if the need arises.
